tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -102,7 +101,7 @@ export class WebGPUBackend extends KernelBackend {
   commandQueue: GPUCommandEncoder[];
 
   private commandQueueOwnedIds = new WeakSet<DataId>();
-  private binaryCache: {[key: string]: WebGPUBinary};
+  private programCache: Map<number, webgpu_program.WebGPUProgram> = new Map();
```

Comment:
**I like this wonderful idea, and I tried to make it work on posenet demo. (code [here](https://github.com/haoyunfeix/tfjs/compare/cache_programs...haoyunfeix:exp_bigint))
But result shows BigInt doesn't perform well, because the arithmetic operations involving BigInt are SLOW.**

![getBigInt](https://user-images.githubusercontent.com/7088194/90586658-4d245b00-e20a-11ea-9a47-5d9019ade781.png)

> The floating-point arithmetic is implemented in hardware, Hence it is faster. The Arbitrary-precision arithmetic is entirely implemented in software hence slower. --[tektutorialshub](https://www.tektutorialshub.com/typescript/typescript-bigint-vs-number/)

Reply:
Hm, I would kind of expect BigInt could perform a little better. If `BigInt(0)` is really slow that's a red flag. BigInts theoretically need allocations (since they can be arbitrarily sized), but up to at least 64 bits ought to be implemented without an allocation (similar to std::string, which can hold short strings internally without any allocations).

That said, fundamentally BigInt and string are the same - they're both an arbitrary length vector of bytes and both could do that internalization optimization. So I'm not too surprised that the performance is similar.

Aside: The simplification that bigint arithmetic is "implemented in software" is kind of misleading. Yes, a floating point operation only requires one instruction, but bigint arithmetic is implemented on top of hardware instructions too, there are just more instructions required. I guess the dealbreaker is that too many instructions are required.

Reply:
@kainino0x I found `BigInt(0)` will take more time than `0n` , e.g. `0 + 0` vs `0n + 0n` vs `BigInt(0) + BigInt(0)` => `3ms` vs `5ms` vs `100ms` in 1 million loops. Bottleneck maybe at BigInt()? I'll continue investigate this and try to make `0n` works on posenet demo to see if we could get some expected data.